### DISCUSSION TITLE
How to correctly simulate `atomicAdd` on `u64` by using two `u32` buffers?

### DISCUSSION THREAD

**Discussion by reem-codes** [gpuweb/gpuweb#4747] [Q&A] [answered]
Hello,

I'm trying to do atomic operations on `u64`. But since it's not supported, the number is stored in TWO `u32` buffers

The issue is that I'm not sure how to do `atomicAdd` correctly to simulate the effect it would have had on `u64`. All while avoiding memory modification by other threads between loading and storing the values.

my current idea is this:

```wgsl

fn tou64(value: u32) -> vec2u {
        return vec2u(u32(value / BASE), value % BASE);
}

fn add(a: vec2u, b: vec2u) -> vec2u {
    let x = a.x + b.x + u32((a.y + b.y) / BASE);
    let y = (a.y + b.y) % BASE;
    return vec2u(x, y);
}

fn main() {
// .....

// convert the value from u32 to 2-buffer representation of u64
let b: vec2u = tou64(value);
// fetch the old value from the 2 buffers
var a = vec2u(0); 
a.x = atomicLoad(&buffer[index]);
a.y = atomicLoad(&buffer[index+1]);
// add the value to the buffer value
let result = add(a, b);
// store back the buffer results 
atomicStore(&buffer[index], result.x);
atomicStore(&buffer[index+1], result.y);
}
```

This works only when no other thread is modifying the buffer at the same `index`. But it's a very weak implementation otherwise. Thread 1 could change the value of `buffer[index+1]` while thread 2 just read the old `buffer[index]` value and the new `buffer[index+1]` value

**Comment by reem-codes** [comment, marked as answer]
Here is the solution for `atomicAddU64`, thanks to [PeterCordes](https://stackoverflow.com/users/224132/peter-cordes) from [stackoverflow](https://stackoverflow.com/a/78763048/8754028)!

 
```wgsl
fn main() {
    // .....

    // convert the value from u32 or f32 to 2-buffer representation of u64
    let b: vec2u = tou64(value);

    // low: no need for modulu since it will wrap by itself
    // take old value on the buffer to check for carry
    let oldValue = atomicAdd(&buffer[low_index], b.y);
    // high: add high part + carry
    // check if the sum cause value to wrap
    atomicAdd(&buffer[high_index], b.x + u32((oldValue + b.y) < b.y));
}
```

To convert u64 to vec2u and back:

```wgsl
// this is just a pseudocode! wgsl doesn't support u64 yet
// do this operation in c++/js or while data processing
fn u64ToVec2u(value: u64) -> vec2u {
    let low = u32(value);
    let high = u32(value >> 32);
    return vec2u(high, low);
}

fn vec2uToU64(value: vec2u) -> u64 {
    return (value.x << 32) + value.y; // x = high, y = low
}

```

and here's a solution for float computations as well, can be done in wgsl

```wgsl
// to save 2 ^ 10 = 1024 (ie 3 digits after the decimal point)
const DEGREE_TO_SAVE = 10; 
fn tou64(value: f32) -> vec2u {
    // modulu is important here!! 
    // because converting from float to u32 will not automatically wrap
    let low = u32((value * pow(2., DEGREE_TO_SAVE)) % pow(2., 32));
    let high = u32(value /  pow(2., 32 - DEGREE_TO_SAVE));
    return vec2u(high, low);
}

fn tof32(value: vec2u) -> f32 {
    return f32(value.x) * pow(2., 32 - DEGREE_TO_SAVE) + f32(value.y) / pow(2., DEGREE_TO_SAVE);

}
```

**Reply by Kangz** [reply]
While this atomicAdd seems like it will work to just have an add-only counter, I don't think it can produce and u64 values back (or `vec2u`) that represents the value in the counter prior to the atomic add.

**Reply by reem-codes** [reply]
True. It's not a correct `atomicAdd` implementation that follows the guidelines because it doesn't return the old value of vec2u. 

Thankfully, I don't need the old value. So the solution above works if the old vec2u value isn't needed 

I hope that future wgsl releases will support more types like `u64`, `f64`, `u16` and their `atomicAdd` implemented
